libazureinit/src/health.rs

```diff
@@ -0,0 +1,418 @@
+) -> Result<(), Error> {
+    _report(
+        ProvisioningState::InProgress,
+        Some(ProvisioningSubStatus::ProvisioningInProgress),
+        Some(message),
+        &config.wireserver,
+    )
+    .await
+}
+
+/// Internal helper: all of the header setup, JSON‐body construction, retry loop, etc.
+#[instrument(err, skip_all)]
+async fn _report(
+    state: ProvisioningState,
+    substatus: Option<ProvisioningSubStatus>,
+    description: Option<&str>,
+    cfg: &crate::config::Wireserver,
+) -> Result<(), Error> {
+    let body = if let Some(sub) = substatus {
+        json!({
+            "state": state.to_string(),
+            "details": {
+                "subStatus": sub.to_string(),
+                "description": description
+            }
+        })
+        .to_string()
+    } else {
+        json!({ "state": state.to_string() }).to_string()
+    };
+
+    tracing::debug!(body=%body, "Built provisioning-health JSON");
+
+    let version = env!("CARGO_PKG_VERSION");
+    let mut headers = HeaderMap::new();
+    headers.insert(
+        USER_AGENT,
+        HeaderValue::from_str(&format!("azure-init v{}", version)).unwrap(),
+    );
+    headers.insert(
+        "x-ms-guest-agent-name",
+        HeaderValue::from_str(&format!("azure-init/{}", version)).unwrap(),
+    );
+    headers
+        .insert("content-type", HeaderValue::from_static("application/json"));
+
+    tracing::info!(?headers, "Prepared HTTP headers");
+
+    let connect_timeout = Duration::from_secs_f64(cfg.connection_timeout_secs);
+    let read_timeout = Duration::from_secs_f64(cfg.read_timeout_secs);
+    let retry_for = Duration::from_secs_f64(cfg.total_retry_timeout_secs);
+
+    let client = Client::builder()
+        .connect_timeout(connect_timeout)
+        .timeout(read_timeout)
+        .build()?;
+
+    let mut remaining = retry_for;
+    while !remaining.is_zero() {
+        let (resp, new_remaining) = http::post(
+            &client,
+            headers.clone(),
+            body.clone(),
+            read_timeout,
+            connect_timeout,
+            remaining,
+            &cfg.health_endpoint,
+        )
+        .await?;
+
+        let status = resp.status();
+        for (key, value) in resp.headers().iter() {
+            tracing::info!(
+                target: "libazureinit::health::status",
+                header = %key,
+                value = ?value,
+                "Wireserver response header"
+            );
+        }
+        tracing::info!(
+            target: "libazureinit::health::status",
+            "Wireserver replied with status {}",
+            status
+        );
+
+        if status.is_success() {
+            tracing::info!(
+                target: "libazureinit::health::status",
+                "Report '{}' succeeded",
+                state
+            );
+            return Ok(());
+        }
+
+        if status == StatusCode::TOO_MANY_REQUESTS
+            || status == StatusCode::SERVICE_UNAVAILABLE
```

Comment:
It might not hurt to retry on `INTERNAL_SERVER_ERROR` as well. I don't know if WireServer returns `INTERNAL_SERVER_ERROR` or any other 5XX codes, but lots of services _do_ even if they don't document that it's a possibility.